File: pmt_solver.py
import time
from decimal import *


class LoanError(Exception):
    
    def __init__(self, message):
        self.message = message

    def __str__(self):
        return __repr__(self.message)



class Loan(object):
    ''' 
    small class to compute solution to PMT on Loan
    construct new Loan giving:
        beginning_value > $0.00,
        future_value >= $0.00, (enter as $0 if fully repaying)
        interest_rate > 0.00, (annual interest rate, 0.05 for 5%)
            if periods are years, enter computed form 0.05
            if periods are months, enter computed form 0.05/12
            if periods are weeks, enter computed form 0.05/52
            if periods are days, computed form enter 0.05/365
        payment_frequency = ["years","months","weeks","days"]
        periods > 0, (enter number of periods)
    '''
    frequencies = {"years": 1, "months": 12, "weeks": 52, "days": 365}
    def __init__(self, beginning_value,
                       future_value,
                       interest_rate,
                       payment_frequency,
                       periods):
        self.beginning_value = beginning_value
        if payment_frequency in self.frequencies:
            self.interest_rate = interest_rate / self.frequencies[payment_frequency]
        else:
            raise LoanError("Incorrect payment frequency!")
        self.periods = periods
        self.future_value = future_value
        self.min_payment = (beginning_value - future_value) / periods
        self.recent_max = 0.0
        self.iterations = 0
        self.solved_payment = 0.0

    def pmt(self):
        ''' main api call to solve for a payment '''
        values = self.set_initial_values()
        self.solve(values)
        print ("Payment: %s\tTotal iterations: %s" % (self.solved_payment, self.iterations))
        return self.solved_payment

    def set_initial_values(self, payment=None):
        ''' compute beginning values for first period value '''
        if not payment:
            ''' set max payment '''
            payment = self.beginning_value * ((1 + self.interest_rate) ** self.periods)
        start_interest = self.beginning_value * self.interest_rate
        end_val = self.beginning_value - (payment - start_interest)
        values = [self.beginning_value, start_interest, payment, end_val]
        return values

    def solve(self, starting_values):
        ''' method implements solution to find payment '''
        self.iterations = self.iterations + 1
        this_payment = starting_values[2]
        loan_schedule = [starting_values]

        for i in range(0,self.periods):
            loan_schedule.append(self.next_period(i, loan_schedule[i]))

        if loan_schedule[self.periods-1][3] > 0.04 or loan_schedule[self.periods-1][3] < -0.04:
            if loan_schedule[self.periods-1][3] < 0.04:
                ''' ending balance negative, payment lower '''
                self.recent_max = this_payment
                next_payment = (this_payment + self.min_payment) / 2
                values = self.set_initial_values(next_payment)
                self.solve(values)
            elif loan_schedule[self.periods-1][3] > -0.04:
                ''' ending balance positive, payment higher '''
                next_payment = (this_payment + self.recent_max) / 2
                values = self.set_initial_values(next_payment)
                self.solve(values)
        else:
            solved_payment = Decimal(this_payment).quantize(Decimal('.01'), rounding=ROUND_UP)
            print ("Solution found! Periodic Payment: %s" % (solved_payment))
            self.solved_payment = float(solved_payment)
            return solved_payment

    def next_period(self, period, period_values):
        # beginning_value = Decimal(period_values[0]).quantize(Decimal('.01'), rounding=ROUND_UP)
        # interest_paid = Decimal(period_values[1]).quantize(Decimal('.01'), rounding=ROUND_UP)
        # payment = Decimal(period_values[2]).quantize(Decimal('.01'), rounding=ROUND_UP)
        # ending_value = Decimal(period_values[3]).quantize(Decimal('.01'), rounding=ROUND_UP)

        # print ("%s\t%s\t\t\t%s\t\t%s\t\t%s" % (period+1, beginning_value, interest_paid,
        #                                    payment, ending_value))


        next_period = [
            period_values[3],
            period_values[3] * self.interest_rate,
            period_values[2],
            period_values[3] - (period_values[2] - (period_values[3] * self.interest_rate))
        ]
        return next_period


if __name__ == "__main__":

    start = time.time()
    new_loan = Loan(100000,0,0.08,"years",25)
    a = new_loan.pmt()
    stop = time.time() - start
    print(stop)

File: ruby/pmt_solver.rb
require "time"

class Loan
    attr_reader :beginning_value, :future_value, 
                :interest_rate, :payment_frequency, :periods,
                :solved_payment, :min_payment, :recent_max, :iterations

    def initialize(args)
        @beginning_value = args[:beginning_value]
        @future_value = args[:future_value]
        ## test for invalid frequency later
        @interest_rate = args[:interest_rate] / frequencies(args[:payment_frequency])
        @payment_frequency = args[:payment_frequency]
        @periods = args[:periods]
        @min_payment = (beginning_value - future_value) / periods
        @recent_max = 0.0
        @iterations = 0
        @solved_payment = 0.0
    end

    def set_initial_values(payment=nil)
        values = Hash.new
        if payment.nil?
            payment = beginning_value * ((1 + interest_rate) ** periods)
        end
        values[:beginning_value] = beginning_value.to_f
        values[:interest_paid] = (beginning_value * interest_rate).to_f
        values[:payment] = payment.to_f
        values[:ending_value] = (beginning_value - (payment - values[:interest_paid])).to_f
        return values
    end

    def frequencies(frequency)
        frequencies = Hash.new
        frequencies = {"years" => 1,
                       "months" => 12,
                       "weeks" => 52,
                       "days" => 365}
        frequencies.fetch(frequency)
    end

    def pmt
        values = []
        starting_value = set_initial_values()
        values << starting_value
        solve(values)
        return solved_payment
    end

    def solve(values)
        @iterations = iterations + 1
        current_payment = values[0][:payment]

        period = 0
        (periods-1).times do
            values << next_period(values[period])
            period = period + 1
        end
        
        if values[-1][:ending_value] > 0.04 or values[-1][:ending_value] < -0.04
            
            # puts "payment #{values[-1][:payment]} ending #{values[-1][:ending_value]}"
            if values[-1][:ending_value] < 0.04
                ## ending balance negative payment smaller
                @recent_max = current_payment
                # puts "recent max now #{recent_max}"
                next_payment = ((current_payment + min_payment) / 2).to_f
                next_values = []
                starting_value = set_initial_values(next_payment)
                next_values << starting_value
                solve(next_values)
            elsif values[-1][:ending_value] > -0.04
                ## ending balance positive, payment bigger
                next_payment = ((current_payment + recent_max) / 2).to_f
                next_values = []
                starting_value = set_initial_values(next_payment)
                next_values << starting_value
                solve(next_values)
            end
        else
            @solved_payment = current_payment.to_f.round(2)
        end
    end

    def next_period(values)
        next_values = {}
        next_values[:beginning_value] = values[:ending_value].to_f
        next_values[:interest_paid] = (next_values[:beginning_value] * interest_rate).to_f
        next_values[:payment] = values[:payment]
        next_values[:ending_value] = (next_values[:beginning_value] - (values[:payment] - next_values[:interest_paid])).to_f
        return next_values
    end
end

if __FILE__ == $0

    loan_book = []
    loan_book << Loan.new({ "beginning_value": 100000,
                            "future_value": 0,
                            "interest_rate": 0.08,
                            "payment_frequency": "years",
                            "periods": 25})
    loan_book << Loan.new({ "beginning_value": 100000,
                            "future_value": 0,
                            "interest_rate": 0.08,
                            "payment_frequency": "months",
                            "periods": 25*12})
    loan_book << Loan.new({ "beginning_value": 100000,
                            "future_value": 0,
                            "interest_rate": 0.08,
                            "payment_frequency": "weeks",
                            "periods": 25*52})
    loan_book << Loan.new({ "beginning_value": 100000,
                            "future_value": 0,
                            "interest_rate": 0.08,
                            "payment_frequency": "days",
                            "periods": 25*365})

    times = Hash.new
    loan_book.each do |loan|
        # puts loan.pmt
        total_time = 0.0
        payment = 0.0
        loops = 20
        loops.times do
            t = Time.new.to_f
            pmt = loan.pmt
            runtime = Time.new.to_f - t
            total_time = total_time + runtime
            payment = pmt
        end
        runtime = total_time / loops
        pmt_info = Hash.new
        pmt_info[:payment] = payment
        pmt_info[:runtime] = runtime
        times[loan.payment_frequency] = pmt_info
    end

    puts times
        
end

